### ISSUE TITLE
Vim9: translation of "s:" may use wrong script ID when calling an imported function

### ISSUE THREAD

**Issue by lacygoill** [vim/vim#6644]
**Describe the bug**

In Vim9 script, the translation of `s:` may use the wrong script ID when calling an imported function from an autoload script which has been manually sourced.

**To Reproduce**

Run this shell command:

    vim -Nu <(cat <<'EOF'
        vim9script
        mkdir('/tmp/some/autoload', 'p')
        mkdir('/tmp/some/import', 'p')

        let lines =<< trim END
            if exists('g:guard') | finish | endif
            let g:guard = 1
            import This from 'that.vim'
            fu some#func()
                call s:This()
            endfu
            call some#func()
        END
        writefile(lines, '/tmp/some/autoload/some.vim')

        lines =<< trim END
            vim9script
            export def This()
            enddef
        END
        writefile(lines, '/tmp/some/import/that.vim')

        set rtp^=/tmp/some
        so /tmp/some/autoload/some.vim
    EOF
    )

`E117` is raised:

    E117: Unknown function: <SNR>2_This

**Expected behavior**

No error is raised, because when Vim calls `s:This()` from `some#func()`, it correctly translates `s:` into `<SNR>3_`, and not into `<SNR>2_`.  `3` is the ID of the script from which `s:This()` is imported; `2` is the ID of the autoload script.

**Environment**

 - Vim version: 8.2 Included patches: 1-1382
 - OS: Ubuntu 16.04.6 LTS
 - Terminal: XTerm(358)

**Additional context**

The issue disappears when the guard in the autoload script is removed:
```vim
vim9script
mkdir('/tmp/some/autoload', 'p')
mkdir('/tmp/some/import', 'p')

let lines =<< trim END
    import This from 'that.vim'
    fu some#func()
        call s:This()
    endfu
    call some#func()
END
writefile(lines, '/tmp/some/autoload/some.vim')

lines =<< trim END
    vim9script
    export def This()
    enddef
END
writefile(lines, '/tmp/some/import/that.vim')

set rtp^=/tmp/some
so /tmp/some/autoload/some.vim
```
It also disappears when the autoload script is not sourced manually:
```vim
vim9script
mkdir('/tmp/some/autoload', 'p')
mkdir('/tmp/some/import', 'p')

let lines =<< trim END
    if exists('g:guard') | finish | endif
    let g:guard = 1
    import This from 'that.vim'
    fu some#func()
        call s:This()
    endfu
    call some#func()
END
writefile(lines, '/tmp/some/autoload/some.vim')

lines =<< trim END
    vim9script
    export def This()
    enddef
END
writefile(lines, '/tmp/some/import/that.vim')

set rtp^=/tmp/some
```
It also disappears when the autoload script is rewritten in Vim9 script:
```vim
vim9script
mkdir('/tmp/some/autoload', 'p')
mkdir('/tmp/some/import', 'p')

let lines =<< trim END
    vim9script
    if exists('g:guard') | finish | endif
    g:guard = 1
    import This from 'that.vim'
    fu some#func()
        call s:This()
    endfu
    call some#func()
END
writefile(lines, '/tmp/some/autoload/some.vim')

lines =<< trim END
    vim9script
    export def This()
    enddef
END
writefile(lines, '/tmp/some/import/that.vim')

set rtp^=/tmp/some
so /tmp/some/autoload/some.vim
```
And it disappears when the autoload function is renamed into a regular global function (`some#func()` → `Func()`):
```vim
vim9script
mkdir('/tmp/some/autoload', 'p')
mkdir('/tmp/some/import', 'p')

let lines =<< trim END
    if exists('g:guard') | finish | endif
    let g:guard = 1
    import This from 'that.vim'
    fu Func()
        call s:This()
    endfu
    call Func()
END
writefile(lines, '/tmp/some/autoload/some.vim')

lines =<< trim END
    vim9script
    export def This()
    enddef
END
writefile(lines, '/tmp/some/import/that.vim')

set rtp^=/tmp/some
so /tmp/some/autoload/some.vim
```
---

I don't know whether this is:

   - a bug which warrants a fix
   - an edge-case which does not warrant a fix
   - an edge-case caused by a bug which warrants a fix

**Comment by brammool** [comment]
Bit of a complicated example.  The first thing that's wrong is using :import in a legacy script.  I'll add an error for that.
Since it works when adding "vim9script" I'll assume that solves the problem.  If there is another one, please create a new issue.

**Comment by brammool** [comment]
Hmm, the docs actually imply that using :import in a legacy Vim script is supposed to work.  I'll have to think about that.

**Comment by lacygoill** [comment]
> Bit of a complicated example.

Sorry, the original example was more complicated, and I simplified it as much as I could.

> The first thing that's wrong is using :import in a legacy script. I'll add an error for that.  Since it works when adding "vim9script" I'll assume that solves the problem. If there is another one, please create a new issue.

I found another example where Vim wrongly translates `s:`.  It's simpler, and this time, adding `vim9script` doesn't fix the issue (so I don't know whether it has the same cause):
```vim
vim9script
mkdir('/tmp/some/plugin', 'p')
mkdir('/tmp/some/import', 'p')
let lines =<< trim END
    vim9script
    import Func from 'some.vim'
    nno <c-b> :echo <sid>Func()<cr>
END
writefile(lines, '/tmp/some/plugin/some.vim')
lines =<< trim END
    vim9script
    export def Func()
    enddef
END
writefile(lines, '/tmp/some/import/some.vim')
set rtp^=/tmp/some
feedkeys("\<c-b>")
```

This code raises `E117`:

    E117: Unknown function: <SNR>2_Func

Again, `<SID>` is translated with a wrong script ID.  `2` is the ID of the plugin script (`plugin/some.vim`).  But `Func()` was imported from `import/some.vim` whose ID is `3`.

**Comment by lacygoill** [comment]
Actually, I think it's a different issue, because it doesn't deal with `s:` but with the pseudo-key `<SID>`.  I guess we can't call a script-local function from a mapping if it has been imported.  Maybe it should be documented at `:h vim9-gotchas`.

**Comment by lacygoill** [comment]
> I guess we can't call a script-local function from a mapping if it has been imported. Maybe it should be documented at :h vim9-gotchas.

As a workaround, we can use a wrapper function.

NOT working:

    import Func from 'some.vim'
    nno <c-b> :call <sid>Func()<cr>
                    ^-------^
                        ✘

Working:

    import Func from 'some.vim'
    nno <c-b> :call <sid>Wrapper()<cr>
    fu s:Wrapper()
        call s:Func()
    endfu

Using the previous example, it gives:
```vim
vim9script
mkdir('/tmp/some/plugin', 'p')
mkdir('/tmp/some/import', 'p')
let lines =<< trim END
    vim9script
    import Func from 'some.vim'
    nno <c-b> :call <sid>Wrapper()<cr>
    fu s:Wrapper()
        call s:Func()
    endfu
END
writefile(lines, '/tmp/some/plugin/some.vim')
lines =<< trim END
    vim9script
    export def Func()
        echo 'test'
    enddef
END
writefile(lines, '/tmp/some/import/some.vim')
set rtp^=/tmp/some
feedkeys("\<c-b>")
```

**Comment by lacygoill** [comment]
As a suggestion, here is a patch to update the documentation:

```diff
diff --git a/runtime/doc/vim9.txt b/runtime/doc/vim9.txt
index 7f7b15bb2..557e2fb4d 100644
--- a/runtime/doc/vim9.txt
+++ b/runtime/doc/vim9.txt
@@ -457,6 +457,19 @@ Some Ex commands can be confused with assignments in Vim9 script: >
 	g:pattern:cmd	# invalid command - ERROR
 	:g:pattern:cmd	# :global command
 
+An imported function cannot be called from a mapping:
+
+	import MyFunc from "thatscript.vim"
+	nnoremap <C-B> :call <SID>MyFunc()<CR>	# ERROR
+
+Use a wrapper function:
+
+	import MyFunc from "thatscript.vim"
+	nnoremap <C-B> :call <SID>MyWrapperFunc('my', 'arguments')<CR>
+	func s:MyWrapperFunc(...)
+	  call call('s:MyFunc', a:000)
+	endfunc
+
 Functions defined with `:def` compile the whole function.  Legacy functions
 can bail out, and the following lines are not parsed: >
 	func Maybe()
```

**Comment by brammool** [comment]
That is a different issue.  I tracked down the original problem: it is caused by sourcing the autoload script directly, combined with using something defined in the script, and bailing out if it's loaded again.  What happens is that calling some#func() looks for the function in the "some.vim" autoload script.  Since this wasn't auto-loaded before, it does that now.  But because of the guard the second time it bails out, and clears out the imports.  Without Vim9 script it would not matter much, but with Vim9 script the script-local items are cleared when sourcing the same script again. 
Best way would be to avoid sourcing the autoload script twice.

**Comment by lacygoill** [comment]
Ok, so it's probably a corner case which doesn't warrant a fix.  Thank you for the explanation.

**Comment by lacygoill** [comment]
Fixed by [8.2.1398](https://github.com/vim/vim/commit/daa2f36573db3e1df7eb1fdbc3a09a2815644048).

This is much better now.  In the past, when calling an autoload function, Vim could source an autoload script even when the function was already defined, which was not documented:

    vim -Nu <(cat <<'EOF'
        call mkdir('/tmp/some/plugin', 'p')
        call mkdir('/tmp/some/autoload', 'p')
        let lines =<< trim END
            let g:counter = get(g:, 'counter', 0) + 1
            fu some#func()
            endfu
        END
        call writefile(lines, '/tmp/some/autoload/some.vim')
        let lines =<< trim END
            nno <c-b> :call some#func()<cr>
        END
        call writefile(lines, '/tmp/some/plugin/some.vim')
        set rtp^=/tmp/some
        so /tmp/some/autoload/some.vim
        au VimEnter * call feedkeys("\<c-b>", 'xt') | echo g:counter
    EOF
    )

    2

Now Vim sources an autoload script only if it's really needed.  Thank you again.

**Comment by brammool** [comment]
When using an imported function, the <SID> will now be used, instead of always looking in the current script.  That's with patch 8.2.1399.
